File: backend-nestjs/apps/patient/src/patient-family-link/patient-family-link.service.ts
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { In, Repository } from 'typeorm';
import { PatientFamilyLinkEntity } from './patient-family-link.entity.js';
import { IdentityPayload, nextId } from '@cdm/shared';

@Injectable()
export class PatientFamilyLinkService {
  constructor(
    @InjectRepository(PatientFamilyLinkEntity)
    private readonly repo: Repository<PatientFamilyLinkEntity>,
  ) {}

  async findAllForPatient(identity: IdentityPayload, patientId: string) {
    return this.repo.find({
      where: {
        tenantId: identity.tenantId,
        orgId: In(identity.allowedOrgIds),
        patientId,
      },
    });
  }

  async linkFamily(identity: IdentityPayload, patientId: string, data: { familyUserId: string, relationship: string }) {
    const entity = this.repo.create({
      id: nextId(),
      tenantId: identity.tenantId,
      orgId: identity.orgId,
      patientId,
      familyUserId: data.familyUserId,
      relationship: data.relationship,
      createdAt: new Date(),
    });
    return this.repo.save(entity);
  }
}

File: backend-java/patient-service/src/main/java/com/cdm/patient/service/PatientFamilyLinkService.java
package com.cdm.patient.service;

import com.cdm.common.security.IdentityPayload;
import com.cdm.patient.entity.PatientFamilyLinkEntity;
import com.cdm.patient.repository.PatientFamilyLinkRepository;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;

@Service
public class PatientFamilyLinkService {
    private final PatientFamilyLinkRepository repository;

    public PatientFamilyLinkService(PatientFamilyLinkRepository repository) {
        this.repository = repository;
    }

    public List<PatientFamilyLinkEntity> findAllForPatient(IdentityPayload identity, Long patientId) {
        return repository.findAllByContextAndPatient(identity.getTenantId(), identity.getAllowedOrgIds(), patientId);
    }

    public PatientFamilyLinkEntity linkFamily(IdentityPayload identity, Long patientId, Long familyUserId, String relationship) {
        PatientFamilyLinkEntity entity = new PatientFamilyLinkEntity();
        entity.setId(com.cdm.common.util.SnowflakeIdGenerator.nextId());
        entity.setTenantId(identity.getTenantId());
        entity.setOrgId(identity.getOrgId());
        entity.setPatientId(patientId);
        entity.setFamilyUserId(familyUserId);
        entity.setRelationship(relationship);
        entity.setCreatedAt(LocalDateTime.now());
        
        return repository.save(entity);
    }
}
